Strategy

Y-Pazniak opened the pull request.

MaximPilipeiko commented on Singletons/OnDemandHolderSingleton.java:
```diff
@@ -0,0 +1,13 @@
+//this is the best solution
+//class creation in java is thread-safety, that means when class creates with an object, it is 100% thread-safe
+public class OnDemandHolderSingleton {
+    public static class SingletonHolder {
```
I think, better to use private to limit access options.
```suggestion
    private static class SingletonHolder {
```

MaximPilipeiko commented on Singletons/OnDemandHolderSingleton.java:
```diff
@@ -0,0 +1,13 @@
+//class creation in java is thread-safety, that means when class creates with an object, it is 100% thread-safe
+public class OnDemandHolderSingleton {
+    public static class SingletonHolder {
+        public static final SingletonHolder HOLDER_INSTANCE = new SingletonHolder();
```
```suggestion
       static final SingletonHolder HOLDER_INSTANCE = new SingletonHolder();
```
interesting discussion about on demand holder idiom
https://stackoverflow.com/questions/21604243/correct-implementation-of-initialization-on-demand-holder-idiom

Y-Pazniak committed:
test

Y-Pazniak committed:
test UI

Y-Pazniak committed:
Strategy realization

MaximPilipeiko commented on .idea/misc.xml:
.idea files should not be commited. Please, update gitignore file.

Y-Pazniak replied:
Fixed. I have problems with my idea after update. Will send you details in TG - probably, need help

MaximPilipeiko commented on src/main/java/org/example/BartenderContext.java:
```diff
@@ -0,0 +1,20 @@
+    private final Map<VisitorType, DrinkStrategy> strategyMap;
+
+    public BartenderContext() {
+        strategyMap = new HashMap<>();
```
EnamMap suits here perfectly.

Y-Pazniak replied:
Replaced

MaximPilipeiko reviewed (changes requested):
Other concerns are sent in Telegram.

Y-Pazniak committed:
HashMap replaced by EnumMap
Strategy choosing replaced to client's side
No data for strategy for now - later
